c10/core/TensorImpl.cpp

```diff
@@ -184,6 +186,35 @@ TensorImpl::TensorImpl(
   // Caffe2 operators create Storages with default devices.
 }
 
+void TensorImpl::_change_backend_component_keys(c10::Device device) {
+  DispatchKey device_key =
+      computeDispatchKey(c10::nullopt, c10::nullopt, device);
+  BackendComponent new_backend =
+      c10::DispatchKeySet(device_key).highestBackendKey();
+  BackendComponent old_backend = key_set_.highestBackendKey();
+
+  // following logic TensorImpl::TensorImpl, update the BackendComponent related
+  // keys to correspond to device
+
+  auto key_set =
+      key_set_ - c10::getAutocastRelatedKeySetFromBackend(old_backend);
+  key_set = key_set | c10::getAutocastRelatedKeySetFromBackend(new_backend);
+
+  // The AutogradRelatedKey will not be present if in inference mode
+  // Only swap in the new AutogradRelatedKey if the previous backend related key
+  // is present
+  auto old_autograd_key = c10::getAutogradRelatedKeySetFromBackend(old_backend);
+  if (key_set.has_all(old_autograd_key)) {
+    key_set = key_set - old_autograd_key;
+    key_set = key_set | c10::getAutogradRelatedKeySetFromBackend(new_backend);
+  }
+
+  // See note [Removing keys from DispatchKeySet Only Affects Functionality
+  // Keys]
+  key_set = key_set.remove_backend(old_backend);
+  key_set_ = key_set | DispatchKeySet(device_key);
+}
```

Comment:
@bdhirsh can you review this part?

Reply:
In particular, the getAutogradRelatedKeySetFromBackend feels due for a refactor; shouldn't this be as simple as just writing out the BC bit in the key set? Not sure why there is so much work that has to be done here.

Reply:
I think you're right that all of the `getAutogradRelatedKeySetFromBackend()` calls above shouldn't be needed. @eellison did you run into issues before adding them?

Reply:
Well... technically, NestedTensor has its own autograd key. I don't think that in practice any of the autograd kernels will impact the result of running meta tensor shape inference though, so maybe... we can just simplify this and not worry about the `AutogradNestedTensor` key here?

Reply:
For my tests you're right I can remove the `getAutogradRelatedKeySetFromBackend` related key logic